socket.go

```diff
@@ -13,13 +13,20 @@ import (
 	"golang.org/x/net/proxy"
 )
 
-type dialer interface {
-	dial(_ context.Context, addr string) (net.Conn, error)
-}
-
-type socket interface {
+// Dialer dials peers using a stream-oriented protocol.
+type Dialer interface {
+	// Dial dials the peer using an implementation defined addr to etablish a
+	// stream-oriented connection with which to make client requests. Dialing
+	// must be cancelable via the provided context.
+	Dial(ctx context.Context, addr string) (net.Conn, error)
+}
+
+// Socket is a generic stream-oriented peer interface that implements
+// net.Listener to handle incoming requests from peers, and Dialer to handler
+// outgoing requests to peers.
+type Socket interface {
```

Comment:
What about splitting this into `Listener` and `Dialer`, there are situations where we want to distinguish them (like proxies), and we can avoid juggling and guessing which is which.